crates/subspace-service/src/sync_from_dsn.rs

```diff
@@ -251,19 +290,70 @@ where
         .saturating_sub(chain_constants.confirmation_depth_k().into());
     let segment_header_downloader = SegmentHeaderDownloader::new(node);
 
-    while let Some(reason) = notifications.next().await {
-        pause_sync.store(true, Ordering::Release);
+    let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+
+    pause_sync.store(true, Ordering::Release);
+
+    let finalized_hash_existed = client.info().finalized_hash != client.info().genesis_hash;
+
+    if fast_sync_enabled && finalized_hash_existed {
+        debug!("Fast sync detected existing finalized hash.");
+    }
+
+    // Run fast-sync first.
+    #[allow(clippy::collapsible_if)] // clippy error: the if statement is not the same
+    if fast_sync_enabled && !finalized_hash_existed {
+        if notifications.next().await.is_some() {
+            let fast_syncer = FastSyncer::new(
+                &segment_headers_store,
+                node,
+                piece_getter,
+                client.clone(),
+                import_queue_service.clone(),
+                network_service.clone(),
+                fast_sync_state,
+                sync_service,
+            );
 
+            let fast_sync_result = fast_syncer.sync().await;
+
+            match fast_sync_result {
+                Ok(fast_sync_result) => {
+                    if fast_sync_result.skipped {
+                        debug!("Fast sync was skipped.");
+                    } else {
+                        last_processed_block_number = fast_sync_result.last_imported_block_number;
+                        last_processed_segment_index = fast_sync_result.last_imported_segment_index;
+                        reconstructor = fast_sync_result.reconstructor;
+
+                        debug!(%last_processed_block_number, %last_processed_segment_index, "Fast sync finished.");
+                    }
+                }
+                Err(err) => {
+                    error!("Fast sync failed: {err}");
+                }
+            }
+
+            notifications_sender
+                .try_send(NotificationReason::WentOnlineSubspace)
+                .map_err(|_| {
+                    sc_service::Error::Other("Can't send sync notification reason.".into())
+                })?;
+        }
+    }
+
+    while let Some(reason) = notifications.next().await {
```

Comment:
You have removed sync pausing when notification is received, this is incorrect and will cause issues